### PR TITLE
Improve key/cert search performance: use id/label to enumerate

### PR DESCRIPTION
**Pull request by harshalgohel**
Add PKCS11_enumerate_*_ext functions to find keys with id or label if supplied.
This change also retains previous behavior of fetching all the keys/certs
if no label or id is supplied.

Add example and test for PKCS11_enumerate_*_ext functions.

If PKCS11_enumerate_*_ext functions are called with NULL template,
behavior is same as before - enumerate everything.

Previous attempt to solve this was in 206af152fbcb48d4149097ed140e4840febd72d8
This improves on it and restores caching to avoid segmentation fault.
And improve performance for further calls.

### CODE REVIEW

**Comment by harshalgohel** [comment]
Hi @mtrojnar 

When used with OpenSSL, it was taking 10 seconds to enumerate over all keys and certs on my HSM.
So this is an attempt to fix that delay, now it only takes 1 second

I would like to contribute these improvements and looking forward to see them in next libp11 release.

I would appreciate any feedback or improvement suggestions on the PR

File: examples/listkeys_ext.c [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+		*op_len = 0;
+		return -1;
+	}
+
+	*op_len = byte_len;
+	while(byte_len--) {
+		/* start parsing from end of hexstring to beginning of hex string */
+		/* len is including '\0' so use pre-decrement */
+		int lsb = getbin(str+--len); // this never goes out of bounds, we will have atleast one byte to process!
+		int msb = str+--len >= str ? getbin(str+len) : 0; // avoid underflow on str (when len is not even we assume 0)
```
[the remark this answers is not in the document]
Comment (harshalgohel) [reply examples/listkeys_ext.c]: Not using pointer arithmatic in latest push.
Add test case to test behavior with even and odd length id string

**Comment by harshalgohel** [comment]
Hey @mtrojnar Can you take a look at updated PR? 

Whenever we provision a certificate in SoftHSM (libp11 test), id and label of corresponding private key is same as the certificate. So I made an assumption that it will be same, and I included label and id both as part of search parameter.

This behavior is not observed in another HSM, id of certificate and private key is same but labels are different.

I could not find any requirement for labels to be same.
So update the patch to remove label as search parameter from find_key(cert) and find_certificate(key) function definitions.

**Comment by mtrojnar** [comment]
This PR seems to break macOS tests.  Can you take a look?

**Comment by harshalgohel** [comment]
Hello @mtrojnar 

I checked on macOS VM (BigSur) and macbook pro (Monterey) and observed something strange

| macOS | libp11 | OpenSSL | segfault | error |
| ----- | -------|---------|----------|-------|
| BigSur | libp11:master | 3 | no | OpenSSL: error during RSA private key parsing |
| BigSur | harshal:key_cert_perf_improv | 3 | no | OpenSSL: error during RSA private key parsing |
| Monterey | libp11:master | 3 | no | OpenSSL: error during RSA private key parsing |
| Monterey | harshal:key_cert_perf_improv | 3 | yes | yes: segfault |

Now there has to be some bug within libp11 dependencies between BigSur and Monterey. I am not familiar with macOS at all. Can you help me debugging this.
Standard tools like valgrind are not directly available, I followed ci.yml to install dependencies into macOS VM and not able to make any progress pin pointing source of the segfault.

**Comment by mtrojnar** [comment]
Well done.  I like the way you implemented this feature.

**Pull request merged**